### ISSUE TITLE
Minion do not reconnect on error

### ISSUE THREAD

**Issue by githubcdr** [saltstack/salt#39222]
### Description of Issue/Question

I noticed that many minions get in disconnected state, this seem to be the case since upgrade to 2016.11.2. At random but also when restarting the salt-master.

What's weird is the fact that salt-call pillar.items result in valid data on a disconnected minion, the connection to the master is functional. 

I tried to setup the minion with settings below, like keepalive etc. but no result.

I think the main issue here is that minions fail to detect a isolated state.

### Setup

Running Salt 2016.11.2 on CentOS with zeromq

```
ipv6: False
verify_env: True
backup_mode: minion
acceptance_wait_time: 10
random_reauth_delay: 10
auth_tries: 10
auth_safemode: False
log_file: /var/log/salt/minion
log_level: error
log_level_logfile: warning
mine_interval: 240
hash_type: sha256
master_alive_interval: 120
master_tries: -1
ping_interval: 2
restart_on_error: True
cache_jobs: False
tcp_keepalive: True
tcp_keepalive_idle: 60
```

### Steps to Reproduce Issue

restart salt-master

### Versions Report

```
Salt Version:
           Salt: 2016.11.2

Dependency Versions:
           cffi: 0.9.2
       cherrypy: 3.2.2
       dateutil: 2.4.0
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: 1.4.6
         Jinja2: 2.8.1
        libgit2: Not Installed
        libnacl: 1.4.3
       M2Crypto: 0.20.2
           Mako: 1.0.0
   msgpack-pure: 0.1.3
 msgpack-python: 0.4.6
   mysql-python: 1.2.5
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.6.6 (r266:84292, Aug 18 2016, 15:13:37)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.5.0
           RAET: 0.6.5
          smmap: 0.9.0
        timelib: 0.2.4
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: centos 6.8 Final
        machine: x86_64
        release: 2.6.32-642.13.1.el6.x86_64
         system: Linux
        version: CentOS 6.8 Final
```

Is this a bug or am I missing something in the config? What can I do to prevent this issue?

**Comment by gtmanfred** [comment]
I think this may be the same issue as here

https://github.com/saltstack/salt/issues/39200

Do you see the minion process running at all after the upgrade on RHEL 6?

I am seeing the minion process die.  Salt-call starts up a salt-minion process completely seperate of the salt-minion that is running with the daemon, so even if the daemon is running and disconnected, salt-call could still connect.

**Comment by githubcdr** [comment]
I think so, but will have to do some more tests tomorrow. I also noticed that when I issue the following command on a disconnected minion

```
salt-call stateus.master connecteced=True master=yadadaa
```

it brings the minion back online, strange behaviour...

**Comment by githubcdr** [comment]
I increased debug logging and noticed that the (disconnected) minion receives the command from the master but can not return.

```
[TRACE   ] Decoding payload: {'load': 'i\xc.....\xba', 'enc': 'aes'}
[INFO    ] User sudo_admin Executing command status.uptime with jid 20170208163719653605
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20170208163719653605', 'tgt': 'origin1-hls.nl', 'ret': '', 'user': 'sudo_admin', 'arg': [], 'fun': 'status.uptime'}
[TRACE   ] Started JIDs: ['20170208154402176593', '20170208154402186325', '20170208154655174789', '20170208154851970356', '20170208154927899134', '20170208160238883053', '20170208161359984066', '20170208161408041406', '201702081626
12444573', '20170208162620550126', '20170208162649212038', '20170208162657312742', '20170208162726426683', '20170208162734498149', '20170208162802860053', '20170208162810980602', '20170208162841846375', '20170208162849937667', '201
70208162932639921', '20170208162940665207', '20170208162959873669', '20170208163007933151', '20170208163052297915', '20170208163100359025', '20170208163219675331', '20170208163227712325', '20170208163247452483', '201702081632554743
95', '20170208163423528348', '20170208163431599811', '20170208163537288506', '20170208163545354236']
[INFO    ] Starting a new job with PID 27619
```

It seems the minion has too many (queued) jid's, is this normal?

What makes it even stranger is that line below works as expected, the message is sent to the master.

```
salt-call event.send myco/mytag foo=Foo bar=Bar
```

Only a restart helps to reconnect.

**Comment by gtmanfred** [comment]
The fact that salt-call is working doesn't seem strange.  It creates a whole new minion connection that is completely seperate from the running daemon.  If a restart of the salt-minion fixes it, it would stand to reason that a salt-call should work.

**Comment by githubcdr** [comment]
Thanks for explaining, I thought the a salt-call would have used the same zeromq connection.

What bothers me is the fact that the disconnected minion receives the commands from the master but is unable to send a response and the fact that this isolated mode is not taken care of in the minion.

I have done some more testing and tweaked the minion config further, let's see where this goes.

**Comment by gtmanfred** [comment]
Thanks, we appreciate the testing.  Unfortunately I am unable to replicate the minions disconnecting :(

Thanks,
Daniel

**Comment by githubcdr** [comment]
I enabled debug ZeroMQ logging and noticed the following;

```
2017-02-10 13:46:30,991 [salt.transport.zeromq][DEBUG   ][25959] ZeroMQ event: {'endpoint': 'tcp://8.8.8.8:4505', 'event': 512, 'value': 24, 'description': 'EVENT_DISCONNECTED'}
2017-02-10 13:46:30,992 [salt.transport.zeromq][DEBUG   ][25959] ZeroMQ event: {'endpoint': 'tcp://8.8.8.8:4505', 'event': 4, 'value': 39048, 'description': 'EVENT_CONNECT_RETRIED'}
2017-02-10 13:46:43,855 [salt.minion      ][DEBUG   ][25959] Minion of "saltstack" is handling event tag '__master_disconnected'
2017-02-10 13:46:43,879 [salt.minion      ][INFO    ][25959] Connection to master saltstack lost
2017-02-10 13:46:43,881 [salt.minion      ][DEBUG   ][25959] Minion of "saltstack" is handling event tag '/salt/minion/minion_schedule_delete_complete'
2017-02-10 13:47:10,079 [salt.transport.zeromq][DEBUG   ][25959] ZeroMQ event: {'endpoint': 'tcp://8.8.8.8:4505', 'event': 2, 'value': 115, 'description': 'EVENT_CONNECT_DELAYED'}
2017-02-10 13:47:10,080 [salt.transport.zeromq][DEBUG   ][25959] ZeroMQ event: {'endpoint': 'tcp://8.8.8.8:4505', 'event': 1, 'value': 23, 'description': 'EVENT_CONNECTED'}
```

Reading the logs you would expect everything returned to normal, but the minion stays disconnected.

Is there anything else I can do, this issue is a pain since it happens so randomly.

**Comment by githubcdr** [comment]
After some more testing I think this issue is related to #36866, it also seems that Windows minions are affected in my experience.

So far it seems this problem occurs when a minion has _master_alive_interval_ config option set.

**Comment by mattp-** [comment]
I think this should be fixed in I fixed this in #49292

**Comment by rallytime** [comment]
@githubcdr ^
